kratos/utilities/pointer_map_communicator.h

```diff
@@ -0,0 +1,623 @@
+    ///@{
+
+    const int mCurrentRank;
+    const TApplyFunctor& mrApplyFunctor;
+
+    void (ProxyType::*mUpdateMethod)(const GlobalPointer<TPointerDataType>&, const TValueDataType&);
+
+    // here we store this in two vector maps(vector of gps, vector of values) instead of one vector map(key:gp, value: std::vector)
+    // because, when updating
+    //      1. Two vector maps will have to only have to append the two vectors for each non-local update.
+    //          Whenever std::vector reaches the capacity, it is doubled,
+    //          so the cost of resizing for vector appending is in O(log N)
+    //      2. If a single vector map is used, first it has to find the gp in the map [ O(log N) if found, otherwise larger] and then append one vector
+    //          for each non-local update.
+    // The method 2 is little bit more expensive than first one AFAIK. so implemented the vector versions. Suggestions are highly appreciated :)
+    //
+    std::vector<GlobalPointerMapType> mNonLocalGlobalPointerMapsVector;
+    std::vector<DataVectorMapType> mNonLocalDataValueMapsVector;
+
+    TGPMapCommunicator& mrPointerCommunicator;
+
+    ///@}
+    ///@name Private operations
+    ///@{
+
+    /**
+     * @brief Assign local values
+     *
+     * This method assigns local gp values using mrApplyFunctor.
+     * It assumes rGlobalPointer is a local gp always, therefore
+     * no checks are performed.
+     *
+     * @param rGlobalPointer    Local Global pointer of the destination
+     * @param rValue            Value to be used in assigning
+     */
+    void AssignLocalData(
+        const GlobalPointer<TPointerDataType>& rGlobalPointer,
+        const TValueDataType& rValue)
+    {
+        KRATOS_TRY
+
+        KRATOS_DEBUG_ERROR_IF(rGlobalPointer.GetRank() != mCurrentRank)
+            << "Using local global pointer update method with a non-local "
+               "global pointer. [ MyPID = "
+            << mCurrentRank
+            << " GlobalPointerRank = " << rGlobalPointer.GetRank() << " ].\n";
+
+        auto gp_pointer = rGlobalPointer;  // required to remove the const from rGlobalPointer
+        mrApplyFunctor(*gp_pointer, rValue);
+
+        KRATOS_CATCH("");
+    }
+
+
+    /**
+     * @brief Assign local and non-local values
+     *
+     * This method assigns local gp values instantly using mrApplyFunctor.
+     * As for the non-local gps, they are stored to be used for future
+     * communication
+     *
+     * @param rGlobalPointer    Global pointer of the destination
+     * @param rValue            Value to be used in assigning
+     */
+    void AssignLocalAndRemoteData(
+        const GlobalPointer<TPointerDataType>& rGlobalPointer,
+        const TValueDataType& rValue)
+    {
+        const int data_rank = rGlobalPointer.GetRank();
+
+        if (data_rank == mCurrentRank) {
+            AssignLocalData(rGlobalPointer, rValue);
+        } else {
+            const int k = OpenMPUtils::ThisThread();
```

Comment:
ok, but do take in mind that this function will eventually be removed if we switch out from openmp.
Instead of an integer in [0,numthreads] it will then devolve the hashtag that identifies the thread. 
(i don't have a better solution, it is just for you to take it into account)